### PR TITLE
Add some libc function

### PR DESCRIPTION
**Pull request by coolyjg**
Finish some libc implementations. 
### stdlib.h
- `calloc`: ax_malloc + memset
- `realloc` declaration: Previous code lacks of declaration
- `strtol`, `atoll`, `strtoll`, `strtoul`, `strtoull`: Some type conversion functions
- `exit`: call `ax_exit`
- `abs`, `llabs`: Get absolution value
### stdio.h
- Change `stdin`, `stdout`, `stderr` definition to `FILE*`. Accordingly change corresponding `print`-like functions using `stdout->fd` rather than number 1. 
- `fopen`: Since libc are responsible for allocating `FILE*` structure in `fopen`, this function acquires both "fs" and "alloc" feature.
- `fflush`: Change input variable type. 
- `vsnprintf`, `snprintf`, `vsprintf`, `vfprintf`, `sprintf`: Some `print`-like implementation inspired by previous `fprintf` implementation.
### time.h
- `clock_gettime`, `gmtime`, `time`, `local_time`, `gettimeofday`: Some `time` implementation based on `ulib/libax/src/time.rs`
- `nanosleep`: Call sleep API in task. TODO:  task can be early woken up by some signal and return remained time.
### unistd.h
- `usleep`, `sleep`: Call sleep API in task
- `getpid`: Call `current()` if "multitask" feature is on, or return 2 ("main" task Id)

Add some `cbindings` syscall implementation
- `ax_exit`
- `ax_clock_gettime`
- `ax_nanosleep`

### CODE REVIEW

File: ulib/libax/src/time.rs [hunk 1]
```diff
@@ -12,6 +12,11 @@ impl Instant {
         Instant(axhal::time::current_time())
     }
 
+    /// Returns the wrapped duration
```
Comment (equation314) [inline ulib/libax/src/time.rs]: Do not export non-std-like functions.

File: ulib/libax/src/cbindings/time.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+use core::{
+    ffi::{c_int, c_long, c_uint},
+    time::Duration,
+};
+
+use crate::{debug, task::sleep, time::Instant};
+
+use super::ctypes;
+
+const NANO_PER_SEC: u64 = 1_000_000_000;
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]: use `axhal::time::NANOS_PER_SEC`

File: ulib/libax/src/cbindings/time.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+/// Sleep some nanoseconds
+///
+/// TODO: should be woken by signals
+#[no_mangle]
+pub unsafe extern "C" fn ax_nanosleep(
+    req: *const ctypes::timespec,
+    rem: *mut ctypes::timespec,
+) -> c_int {
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs:38]: Check `req`, `rem` validation

File: ulib/libax/src/cbindings/time.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+/// Get clock time like `ax_time`
+#[no_mangle]
+pub unsafe extern "C" fn ax_clock_gettime(tp: *mut ctypes::timespec) -> c_int {
+    let now = Instant::now().get_duration();
+    let ret = ctypes::timespec {
+        tv_sec: now.as_secs() as c_long,
+        tv_nsec: now.as_nanos() as c_long,
+    };
+    unsafe {
+        *tp = ret;
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]: Check `tp` validation

File: ulib/libax/src/cbindings/ctask.rs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+//! This crate includes common task syscall without `multitask` feature
+
+use crate::debug;
+use crate::task::exit;
+use core::ffi::c_int;
+
+/// Exit current task
+#[no_mangle]
+pub unsafe extern "C" fn ax_exit(exit_code: c_int) -> ! {
+    debug!("exit called");
```
Comment (equation314) [inline ulib/libax/src/cbindings/ctask.rs]: Remove the debug info as it has been printed in `exit`

File: ulib/libax/src/cbindings/thread.rs
Comment (equation314) [inline ulib/libax/src/cbindings/thread.rs:1]: Merge `ctask.rs` and `thread.rs`?

File: ulib/c_libax/src/time.c [hunk 6]
```diff
@@ -44,3 +163,14 @@ int utimes(const char *filename, const struct timeval times[2])
...
+// TODO: Should match _clk,
+int clock_gettime(clockid_t _clk, struct timespec *ts)
+{
+    return ax_clock_gettime(ts);
+}
+
+int nanosleep(const struct timespec *req, struct timespec *rem)
+{
+    return ax_nanosleep(req, rem);
+}
```
Comment (equation314) [inline ulib/c_libax/src/time.c]: New line at EOF

File: ulib/libax/src/cbindings/ctask.rs
Comment (equation314) [inline ulib/libax/src/cbindings/ctask.rs:1]: Merge `ctask.rs` and `thread.rs`?

File: ulib/libax/src/cbindings/time.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+#[no_mangle]
+pub unsafe extern "C" fn ax_time() -> c_long {
+    let ret = Instant::now().get_duration().as_secs() as c_long;
+    debug!("ax_time return time: {}", ret);
+    ret
+}
+
+/// Get clock time like `ax_time`
+#[no_mangle]
+pub unsafe extern "C" fn ax_clock_gettime(tp: *mut ctypes::timespec) -> c_int {
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]: `tp` => `ts`

File: ulib/c_libax/src/time.c [hunk 8]
```diff
@@ -1,40 +1,159 @@
...
+    static struct tm tm;
+    return localtime_r(timep, &tm);
 }
 
-// TODO:
 time_t time(time_t *t)
 {
-    unimplemented();
-    return 0;
+    time_t ret = (time_t)ax_time();
```
Comment (equation314) [inline ulib/c_libax/src/time.c]: call `ax_clock_gettime`

File: ulib/libax/src/cbindings/time.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    let now = Instant::now().get_duration();
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]: use `axhal::time::current_time()`

File: ulib/libax/src/cbindings/time.rs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+use super::ctypes;
+
+const NANO_PER_SEC: u64 = 1_000_000_000;
+
+/// Get time
+///
+/// TODO: It counts seconds since booting, so it's not real time
+#[no_mangle]
+pub unsafe extern "C" fn ax_time() -> c_long {
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]: Duplicated with `ax_clock_gettime()`, remove this.

File: ulib/libax/src/cbindings/time.rs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    debug!("ax_clock_gettime: {}s, {}ns", now.as_secs(), now.as_nanos());
+    return 0;
+}
+
+/// Sleep some seconds
+///
+/// TODO: currently assume that thread successfully sleeps for `seconds` time
+#[no_mangle]
+pub unsafe extern "C" fn ax_sleep(seconds: c_uint) -> c_uint {
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]: Duplicated with `ax_nanosleep()`, remove this.

File: ulib/libax/src/cbindings/time.rs [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    sleep(Duration::from_secs(seconds as u64));
+    return 0;
+}
+
+/// Sleep some useconds
+#[no_mangle]
+pub unsafe extern "C" fn ax_usleep(useconds: c_uint) -> c_int {
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]: Duplicated with `ax_nanosleep()`, remove this.

File: ulib/c_libax/src/unistd.c [hunk 13]
```diff
@@ -10,18 +10,24 @@ uid_t geteuid(void)
...
+    return 2;
+#endif
 }
 
-// TODO:
 unsigned int sleep(unsigned int seconds)
 {
-    unimplemented();
-    return 0;
+    return ax_sleep(seconds);
```
Comment (equation314) [inline ulib/c_libax/src/unistd.c]: Call `nanosleep`

File: ulib/c_libax/src/unistd.c [hunk 14]
```diff
@@ -10,18 +10,24 @@ uid_t geteuid(void)
...
+}
+
+int usleep(unsigned int useconds)
+{
```
Comment (equation314) [inline ulib/c_libax/src/unistd.c:37]: Call `nanosleep`

File: ulib/c_libax/src/time.c [hunk 15]
```diff
@@ -1,40 +1,159 @@
...
+#define LEAPOCH       (946684800LL + 86400 * (31 + 29))
+#define DAYS_PER_400Y (365 * 400 + 97)
+#define DAYS_PER_100Y (365 * 100 + 24)
+#define DAYS_PER_4Y   (365 * 4 + 1)
+
 // TODO:
 size_t strftime(char *__restrict__ _Buf, size_t _SizeInBytes, const char *__restrict__ _Format,
                 const struct tm *__restrict__ _Tm)
 {
     return 0;
```
Comment (equation314) [inline ulib/c_libax/src/time.c:25]: `unimplemented()`

File: ulib/c_libax/src/time.c [hunk 16]
```diff
@@ -1,40 +1,159 @@
...
 #include <time.h>
 
+#include <libax.h>
+
+const int SEC_PER_MIN = 60;
+const int SEC_PER_HOUR = 3600;
+const int MIN_PER_HOUR = 60;
+const int HOUR_PER_DAY = 24;
+
+#define LEAPOCH       (946684800LL + 86400 * (31 + 29))
```
Comment (equation314) [inline ulib/c_libax/src/time.c:15]: Add comment: `/* 2000-03-01 (mod 400 year, immediately after feb29 */`

File: ulib/c_libax/src/stdout.c
Comment (equation314) [inline ulib/c_libax/src/stdout.c:1]: Merge into `stdio.c`

File: ulib/c_libax/src/stdin.c
Comment (equation314) [inline ulib/c_libax/src/stdin.c:1]: Merge into `stdio.c`

File: ulib/c_libax/src/stderr.c
Comment (equation314) [inline ulib/c_libax/src/stderr.c:1]: Merge into stdio.c

File: ulib/c_libax/include/stddef.h [hunk 17]
```diff
@@ -7,6 +7,9 @@
 typedef uintptr_t size_t;
 typedef intptr_t ssize_t;
 typedef ssize_t ptrdiff_t;
+typedef int64_t off_t;
```
Comment (equation314) [inline ulib/c_libax/include/stddef.h]: Already defined in `sys/types.h`

File: ulib/c_libax/include/locale.h [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    const struct __locale_map *next;
+};
+
+struct __locale_struct {
+    const struct __locale_map *cat[6];
+};
+
+typedef struct __locale_struct *locale_t;
+
+#endif
```
Comment (equation314) [inline ulib/c_libax/include/locale.h]: Add new line

File: ulib/c_libax/include/limits.h [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+#ifndef NAME_MAX
+#define NAME_MAX 255
+#endif
+#define TZNAME_MAX 6
+
+#define PATH_MAX  4096
+#define SSIZE_MAX LONG_MAX
+#define CHAR_MAX  127
+
+#endif
```
Comment (equation314) [inline ulib/c_libax/include/limits.h]: new line

File: ulib/c_libax/include/stdio.h [hunk 20]
```diff
@@ -1,34 +1,61 @@
...
+
 #define FILENAME_MAX 4096
 
+int getchar();
+int putchar(int);
+int puts(const char *s);
+void fprintf(int f, const char *fmt, ...);
+
+#ifdef AX_CONFIG_ALLOC
+#ifdef AX_CONFIG_FS
```
Comment (equation314) [inline ulib/c_libax/include/stdio.h]: `#if defined(AX_CONFIG_ALLOC) && defined(AX_CONFIG_FS)`

File: ulib/c_libax/src/stdio.c [hunk 21]
```diff
@@ -93,7 +95,7 @@ static void printptr(uint64_t value)
     for (j = 0; j < (sizeof(uint64_t) * 2); j++, value <<= 4)
         buf[i++] = digits[value >> (sizeof(uint64_t) * 8 - 4)];
     buf[i] = 0;
-    out(stdout, buf, i);
+    out(stdout->fd, buf, i);
 }
 
 // int getchar()
```
Comment (equation314) [inline ulib/c_libax/src/stdio.c:101]: Can this be uncommented?

File: ulib/c_libax/src/stdio.c [hunk 22]
```diff
@@ -196,3 +198,280 @@ void fprintf(int f, const char *restrict fmt, ...)
...
+        flags |= O_CREAT;
+    if (*mode == 'w')
+        flags |= O_TRUNC;
+    if (*mode == 'a')
+        flags |= O_APPEND;
+    return flags;
+}
+
+#ifdef AX_CONFIG_ALLOC
+#ifdef AX_CONFIG_FS
```
Comment (equation314) [inline ulib/c_libax/src/stdio.c]: `#if defined(AX_CONFIG_ALLOC) && defined(AX_CONFIG_FS)`

File: ulib/c_libax/src/stdio.c [hunk 23]
```diff
@@ -196,3 +198,280 @@ void fprintf(int f, const char *restrict fmt, ...)
...
+
+    // TODO: Should set errno
+    if (!strchr("rwa", *mode)) {
+        return 0;
+    }
+
+    f = (FILE *)malloc(sizeof(FILE));
+
+    flags = __fmodeflags(mode);
+    int ret = ax_fopen(filename, flags, f);
```
Comment (equation314) [inline ulib/c_libax/src/stdio.c]: Re-implement with `ax_open`

File: ulib/libax/src/cbindings/fs.rs [hunk 24]
```diff
@@ -244,3 +244,30 @@ pub unsafe extern "C" fn ax_getcwd(buf: *mut c_char, size: usize) -> *mut c_char
         }
     })
 }
+
+/// Open a file by `filename` and insert it into the file descriptor table
+///
+/// Return 0 if success
+#[no_mangle]
+pub unsafe extern "C" fn ax_fopen(
```
Comment (equation314) [inline ulib/libax/src/cbindings/fs.rs]: Duplicated with `ax_open`

**Comment by coolyjg** [comment]
- Fixed problems mentioned above
  - remove non-std-like `Instant` function
  - check `time` related variables validation
  - merge `ctask.rs and `thread.rs` to `thread.rs`
  - remove `ax_time`, only reserve `ax_clock_gettime`
  - call `axhal::time::current_time()` in ax_clock_gettime
  - remove `ax_sleep`, `ax_usleep`, reserver `ax_nanosleep`
  - merge `stdin.c`, `stdout.c`, `stderr.c` into `stdio.c`
  - remove `ax_fopen`, and re-implemente `fopen` with `ax_open`
- Add `strtoll`, `strtoul`, `strtoull` implementation inspired by unikraft. 
- `getchar()` remains unsolved.

File: ulib/c_libax/include/ctype.h
Comment (equation314) [inline ulib/c_libax/include/ctype.h:1]: Use `#define` for simple functions?

```c
#define isalpha(a) (((unsigned)(a)|32)-'a') < 26)
// ...
```

File: ulib/c_libax/src/stdio.c [hunk 25]
```diff
@@ -196,3 +214,225 @@ void fprintf(int f, const char *restrict fmt, ...)
...
+
+    // TODO: Should set errno
+    if (!strchr("rwa", *mode)) {
+        return 0;
+    }
+
+    f = (FILE *)malloc(sizeof(FILE));
+
+    flags = __fmodeflags(mode);
+    int fd = ax_open(filename, flags, mode);
```
Comment (equation314) [inline ulib/c_libax/src/stdio.c]: `mode` type in `ax_open` is not `char *`

File: ulib/c_libax/include/time.h [hunk 26]
```diff
@@ -2,9 +2,16 @@
...
 #include <stddef.h>
+#include <sys/time.h>
 
 typedef long time_t;
 
+#define CLOCK_REALTIME  0
+#define CLOCK_MONOTONIC 1
+#define CLOCKS_PER_SEC  1000000L
+
+#define ax_time_usec_to_nsec(usec) ((usec)*1000UL)
```
Comment (equation314) [inline ulib/c_libax/include/time.h]: This definition is useless, use `(usec)*1000` in code directly.

File: ulib/libax/src/cbindings/thread.rs [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+    exit(exit_code)
+}
+
+/// Get current thread ID
+#[cfg(feature = "multitask")]
+#[no_mangle]
+pub unsafe extern "C" fn ax_getpid() -> c_int {
+    ax_call_body!(ax_getpid, {
+        let pid = current().id().as_u64() as c_int;
+        debug!("getpid return {}", pid);
```
Comment (equation314) [inline ulib/libax/src/cbindings/thread.rs]: This debug info is already printed in `ax_call_body`

File: ulib/libax/src/cbindings/thread.rs [hunk 28]
```diff
@@ -0,0 +1,26 @@
+//! This crate includes syscalls that need `multitask` feature
+
+use crate::debug;
+use crate::task::exit;
+use axerrno::{LinuxError, LinuxResult};
```
Comment (equation314) [inline ulib/libax/src/cbindings/thread.rs]: Remove unused imports

File: ulib/libax/src/cbindings/time.rs [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+        }
+        let now = current_time();
+        let ret = ctypes::timespec {
+            tv_sec: now.as_secs() as c_long,
+            tv_nsec: now.subsec_nanos() as c_long,
+        };
+        unsafe {
+            *ts = ret;
+        }
+        debug!("ax_clock_gettime: {}s, {}ns", now.as_secs(), now.as_nanos());
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs]:
```rust
debug!("ax_clock_gettime: {}.{:09}s", ret.tv_sec, ret.tv_nsec);
```

File: ulib/libax/src/cbindings/time.rs [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+pub unsafe extern "C" fn ax_nanosleep(
+    req: *const ctypes::timespec,
+    rem: *mut ctypes::timespec,
+) -> c_int {
+    ax_call_body!(ax_nanosleep, {
+        if req.is_null() || (*req).tv_nsec < 0 || (*req).tv_nsec > 999999999 {
+            return Err(LinuxError::EINVAL);
+        }
+
+        let total_nano = (*req).tv_sec as u64 * NANOS_PER_SEC + (*req).tv_nsec as u64;
```
Comment (equation314) [inline ulib/libax/src/cbindings/time.rs:45]: Add debug info for arguments:


```rust
debug!("ax_nanosleep <= {}.{:09}s", (*req).tv_sec, (*req).tv_nsec);
```

**Comment by chyyuu** [comment]
@coolyjg Very good! try to make your work better.

**Comment by coolyjg** [comment]
This PR will be clearer in #38

**Pull request closed without merging**